File: python/rspm/supervisor.py
"""Detached sidecar supervisor for rspm."""

from dataclasses import dataclass
import os
from pathlib import Path
import subprocess
import time

from rspm.client import RspmClient


@dataclass(frozen=True)
class RspmSupervisor:
    """Ensure a detached ``rspmd`` sidecar is available for a host program.

    :param host: Daemon TCP host.
    :type host: str
    :param port: Daemon TCP port.
    :type port: int
    :param rspm_bin: Path to the ``rspm`` executable.
    :type rspm_bin: str | Path | None
    :param log_dir: Directory for daemon stdout/stderr and task logs.
    :type log_dir: str | Path
    :param state_dir: Directory for pid, applied config, and event state.
    :type state_dir: str | Path
    :param socket_path: Unix socket path used by daemon run arguments.
    :type socket_path: str | Path
    :param token: Optional local auth token.
    :type token: str | None
    :param startup_timeout: Seconds to wait for daemon readiness.
    :type startup_timeout: float
    :param ownership: Sidecar ownership policy. Currently only ``detached`` is supported.
    :type ownership: str
    """

    host: str = "127.0.0.1"
    port: int = 27691
    rspm_bin: str | Path | None = None
    log_dir: str | Path = ".rspm/logs"
    state_dir: str | Path = ".rspm/state"
    socket_path: str | Path = ".rspm/run/rspmd.sock"
    token: str | None = None
    startup_timeout: float = 10.0
    ownership: str = "detached"

    def __post_init__(self) -> None:
        if self.ownership != "detached":
            raise ValueError(f"unsupported ownership [{self.ownership}]")

    @property
    def endpoint(self) -> str:
        """Return the sidecar TCP endpoint."""

        return f"tcp://{self.host}:{self.port}"

    def client(self) -> RspmClient:
        """Create a TCP client for the configured sidecar endpoint."""

        client = RspmClient.connect_tcp(self.host, self.port)
        if self.token is not None:
            client.with_token(self.token)
        return client

    def daemon_command(self, config_path: str | Path) -> list[str]:
        """Build the detached ``rspm daemon run`` command."""

        command = [
            str(self.rspm_bin or os.environ.get("RSPM_BIN", "rspm")),
            "daemon",
            "run",
            str(config_path),
            f"{self.host}:{self.port}",
            str(self.log_dir),
            str(self.state_dir),
            str(self.socket_path),
        ]
        if self.token is not None:
            command.extend(["--token", self.token])
        return command

    def ensure_daemon(self, config_path: str | Path) -> RspmClient:
        """Return a ready client, spawning a detached sidecar if needed."""

        if self._is_ready():
            return self.client()

        self._ensure_config_source(config_path)
        self._spawn_daemon(config_path)
        return self._wait_ready()

    def _is_ready(self) -> bool:
        try:
            self.client().list_tasks()
        except OSError:
            return False
        except RuntimeError:
            return False
        return True

    def _ensure_config_source(self, config_path: str | Path) -> None:
        config = Path(config_path)
        applied_config = Path(self.state_dir) / "applied.toml"
        if config.exists() or applied_config.exists():
            return
        raise FileNotFoundError(
            f"missing config [{config}] and no applied config [{applied_config}]"
        )

    def _spawn_daemon(self, config_path: str | Path) -> None:
        log_dir = Path(self.log_dir)
        state_dir = Path(self.state_dir)
        socket_path = Path(self.socket_path)
        log_dir.mkdir(parents=True, exist_ok=True)
        state_dir.mkdir(parents=True, exist_ok=True)
        if socket_path.parent != Path(""):
            socket_path.parent.mkdir(parents=True, exist_ok=True)

        kwargs: dict[str, object] = {}
        if os.name == "nt":
            kwargs["creationflags"] = (
                subprocess.CREATE_NEW_PROCESS_GROUP | subprocess.DETACHED_PROCESS
            )
        else:
            kwargs["start_new_session"] = True

        with (log_dir / "rspmd.stdout.log").open("ab") as stdout, (
            log_dir / "rspmd.stderr.log"
        ).open("ab") as stderr:
            process = subprocess.Popen(
                self.daemon_command(config_path),
                stdin=subprocess.DEVNULL,
                stdout=stdout,
                stderr=stderr,
                **kwargs,
            )
        (state_dir / "rspmd.pid").write_text(str(process.pid))

    def _wait_ready(self) -> RspmClient:
        deadline = time.monotonic() + self.startup_timeout
        while True:
            if self._is_ready():
                return self.client()
            if time.monotonic() >= deadline:
                raise TimeoutError(f"rspmd did not become ready at [{self.host}:{self.port}]")
            time.sleep(0.1)

File: crates/rspm-sdk/src/supervisor.rs
use std::net::SocketAddr;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

use anyhow::{Context, Result};

use crate::transport::TcpRspmClient;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DaemonOwnership {
    Detached,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DaemonCommandSpec {
    pub program: PathBuf,
    pub args: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct RspmSupervisor {
    addr: SocketAddr,
    binary_path: PathBuf,
    log_dir: PathBuf,
    state_dir: PathBuf,
    socket_path: PathBuf,
    token: Option<String>,
    startup_timeout: Duration,
    ownership: DaemonOwnership,
}

impl Default for RspmSupervisor {
    fn default() -> Self {
        Self::new()
    }
}

impl RspmSupervisor {
    pub fn new() -> Self {
        Self {
            addr: SocketAddr::from(([127, 0, 0, 1], 27691)),
            binary_path: std::env::var_os("RSPM_BIN")
                .map(PathBuf::from)
                .unwrap_or_else(|| PathBuf::from("rspm")),
            log_dir: PathBuf::from(".rspm/logs"),
            state_dir: PathBuf::from(".rspm/state"),
            socket_path: PathBuf::from(".rspm/run/rspmd.sock"),
            token: None,
            startup_timeout: Duration::from_secs(10),
            ownership: DaemonOwnership::Detached,
        }
    }

    pub fn addr(mut self, addr: SocketAddr) -> Self {
        self.addr = addr;
        self
    }

    pub fn binary_path(mut self, path: impl Into<PathBuf>) -> Self {
        self.binary_path = path.into();
        self
    }

    pub fn log_dir(mut self, path: impl Into<PathBuf>) -> Self {
        self.log_dir = path.into();
        self
    }

    pub fn state_dir(mut self, path: impl Into<PathBuf>) -> Self {
        self.state_dir = path.into();
        self
    }

    pub fn socket_path(mut self, path: impl Into<PathBuf>) -> Self {
        self.socket_path = path.into();
        self
    }

    pub fn token(mut self, token: impl Into<String>) -> Self {
        self.token = Some(token.into());
        self
    }

    pub fn startup_timeout(mut self, timeout: Duration) -> Self {
        self.startup_timeout = timeout;
        self
    }

    pub fn ownership(&self) -> DaemonOwnership {
        self.ownership
    }

    pub fn daemon_command_spec(&self, config: impl AsRef<Path>) -> DaemonCommandSpec {
        let mut args = vec![
            "daemon".to_string(),
            "run".to_string(),
            config.as_ref().display().to_string(),
            self.addr.to_string(),
            self.log_dir.display().to_string(),
            self.state_dir.display().to_string(),
            self.socket_path.display().to_string(),
        ];
        if let Some(token) = &self.token {
            args.push("--token".to_string());
            args.push(token.clone());
        }
        DaemonCommandSpec {
            program: self.binary_path.clone(),
            args,
        }
    }

    pub async fn ensure_daemon(&self, config: impl AsRef<Path>) -> Result<TcpRspmClient> {
        if let Ok(client) = self.connect_ready().await {
            return Ok(client);
        }

        self.ensure_config_source(config.as_ref())?;
        self.spawn_daemon(config.as_ref())?;
        self.wait_ready().await
    }

    fn ensure_config_source(&self, config: &Path) -> Result<()> {
        let applied_config = self.state_dir.join("applied.toml");
        if config.exists() || applied_config.exists() {
            return Ok(());
        }
        anyhow::bail!(
            "missing config [{}] and no applied config [{}]; pass an existing config path before spawning rspmd",
            config.display(),
            applied_config.display()
        );
    }

    fn spawn_daemon(&self, config: &Path) -> Result<()> {
        std::fs::create_dir_all(&self.log_dir).with_context(|| {
            format!(
                "failed to create daemon log directory [{}]",
                self.log_dir.display()
            )
        })?;
        std::fs::create_dir_all(&self.state_dir).with_context(|| {
            format!(
                "failed to create daemon state directory [{}]",
                self.state_dir.display()
            )
        })?;
        if let Some(parent) = self.socket_path.parent() {
            std::fs::create_dir_all(parent).with_context(|| {
                format!(
                    "failed to create daemon socket directory [{}]",
                    parent.display()
                )
            })?;
        }

        let stdout_path = self.log_dir.join("rspmd.stdout.log");
        let stderr_path = self.log_dir.join("rspmd.stderr.log");
        let stdout = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&stdout_path)
            .with_context(|| format!("failed to open daemon stdout [{}]", stdout_path.display()))?;
        let stderr = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&stderr_path)
            .with_context(|| format!("failed to open daemon stderr [{}]", stderr_path.display()))?;

        let spec = self.daemon_command_spec(config);
        let mut command = Command::new(&spec.program);
        command
            .args(&spec.args)
            .stdin(Stdio::null())
            .stdout(Stdio::from(stdout))
            .stderr(Stdio::from(stderr));
        detach_daemon_command(&mut command);
        let child = command
            .spawn()
            .with_context(|| format!("failed to spawn rspmd via [{}]", spec.program.display()))?;
        std::fs::write(self.state_dir.join("rspmd.pid"), child.id().to_string()).with_context(
            || {
                format!(
                    "failed to write daemon pid file [{}]",
                    self.state_dir.join("rspmd.pid").display()
                )
            },
        )?;
        Ok(())
    }

    async fn wait_ready(&self) -> Result<TcpRspmClient> {
        let deadline = Instant::now() + self.startup_timeout;
        loop {
            if let Ok(client) = self.connect_ready().await {
                return Ok(client);
            }
            if Instant::now() >= deadline {
                anyhow::bail!("rspmd did not become ready at [{}]", self.addr);
            }
            tokio::time::sleep(Duration::from_millis(100)).await;
        }
    }

    async fn connect_ready(&self) -> Result<TcpRspmClient> {
        let mut client = TcpRspmClient::connect(self.addr).await?;
        if let Some(token) = &self.token {
            client = client.with_token(token.clone());
        }
        let _ = client.list_tasks().await?;
        Ok(client)
    }
}

#[cfg(unix)]
fn detach_daemon_command(command: &mut Command) {
    use std::os::unix::process::CommandExt;

    unsafe {
        command.pre_exec(|| {
            if libc::setsid() == -1 {
                return Err(std::io::Error::last_os_error());
            }
            Ok(())
        });
    }
}

#[cfg(not(any(unix, windows)))]
fn detach_daemon_command(_command: &mut Command) {}

#[cfg(windows)]
fn detach_daemon_command(command: &mut Command) {
    use std::os::windows::process::CommandExt;

    const CREATE_NEW_PROCESS_GROUP: u32 = 0x0000_0200;
    const DETACHED_PROCESS: u32 = 0x0000_0008;

    command.creation_flags(CREATE_NEW_PROCESS_GROUP | DETACHED_PROCESS);
}
